### PR TITLE
DL3-113 Redirect on LTI 1.3 controller that pulls token

### PR DESCRIPTION
**Pull request by amgillenwater**
## Description

This code allows for browsers that block third party cookies to use a cookieless route to complete the OIDC flow. Instead of cookies, this route verifies the state through the stages of the OIDC launch by using postMessages passed between the middleware and the LMS. 

More specifically, this code bypasses some cookie-based nonce checks in various controllers (thanks Leanne!) and also adds a state and nonce check to the frontend before allowing the user to either access content in their LMS or add content through deeplinking (instructors).

### Motivation and Context

Safari and iOS users are unable to use Lumen One because of the LTI1.3 flow that uses cookies. This PR implements the cookie-less LTI1.3 flow. 

### Fixes
[DL3-134](https://lumenlearning.atlassian.net/browse/DL3-113)

## How Has This Been Tested?

-Locally across browsers and on iOS using Canvas and local middleware tool
-Currently testing other LMS. 
-If the state and nonce check does not pass, currently this is failing with no user feedback (just a blank screen presents that does not advance foward) DL3-118 will add in more elegant user feedback to errors.

## Screenshots
Screenshots and screen recordings are not super helpful, but here is a launch on my iPhone using Safari. If you'd like to watch the process but with access to dev tools and some logs to verify the process, happy to host a meeting to show you! Just put time on my calendar or DM me. 

https://user-images.githubusercontent.com/63607748/224076666-69933ac0-63f4-4e8c-b7e6-0840b5b2dce6.mov

---

## Database changes

none

## ⚠️ Deployment instructions ⚠️

**For the Front End React Component**

1. `cd ./src/main/frontend/`
2. then run `npm install` to install styled components dependency

## New ENV variables

N/A

---

## Checklist

- [x] I have reviewed the AC for this feature and my changes meet all requirements
- [ ] (For UI changes) I have considered the accessibility of this feature (see https://www.a11yproject.com/checklist/ for a high-level checklist)
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have requested a review from at least one other dev

[DL3-134]: https://lumenlearning.atlassian.net/browse/DL3-134?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/resources/templates/lti3Redirect.html [hunk 1]
```diff
@@ -22,9 +22,21 @@
...
+    <script th:inline="javascript">
+        let state = "made up string"
+        let nonceStorage = JSON.parse(window.localStorage.getItem('nonce'));
+        let nonce = nonceStorage.value;
+        console.log("nonce from storage: ", nonce)
+        let platformOrigin = [[${ iss }]]
 
-    <script type="text/javascript">
-        document.redirectForm.submit();
+        let motherMayIRedirect = new LtiStorage().validateStateAndNonce(state, nonce, platformOrigin, window)
```
Comment (amgillenwater) [inline src/main/resources/templates/lti3Redirect.html]: Renaming this var is on the to do list also 😅

**Comment by LeftShotDev** [comment]
@smith750 Do we want to confirm which rc branch this work should go in?

**Comment by smith750** [comment]
Good call, @LeftShotDev - this should go into rc/unuk-middleware

File: src/main/frontend/src/app/ltistorage.js [hunk 2]
```diff
@@ -0,0 +1,488 @@
+//  This code originates from 1EdTech's LTI Postmessage Example Library.
```
Comment (smith750) [inline src/main/frontend/src/app/ltistorage.js]: Just out of curiosity, why does this code need to exist twice?

**Review by smith750 — APPROVED** [review]
This looks good - thank you so much Ashley!  We'll figure out the other LMSs

File: src/main/frontend/src/app/ltistorage.js (hunk 2 above)
Comment (amgillenwater) [reply src/main/frontend/src/app/ltistorage.js]: the react app needs access inside its own `/src` file.

Comment (smith750) [reply src/main/frontend/src/app/ltistorage.js]: Ah.  That's lamentable but we can try to think through a fix to that in the future sometime....

**Review by LeftShotDev — APPROVED** [review]
LGTM!

**Review by leannethng — APPROVED** [review]
Thanks Ashley, we are going to merge this into Unuk now! 🎉

**Pull request merged**